cfdm/read_write/netcdf/netcdfwrite.py

```diff
@@ -4595,7 +4657,131 @@ def write(
             # Whether or not to name dimension corodinates in the
             # 'coordinates' attribute
             "coordinates": bool(coordinates),
+            # Dictionary of netCDF variable names and netCDF
+            # dimensions keyed by items of the field (such as a
+            # coordinate or a coordinate reference)
+            "seen": {},
+            # Dry run: populate 'seen' dict without actually writing to file.
+            "dry_run": False,
+            # To indicate if the previous iteration was a dry run:
+            "post_dry_run": False,
+            # Note: need write_vars keys to specify dry runs (iterations)
+            # and subsequent runs despite them being implied by the mode ('r'
+            # and 'a' for dry_run and post_dry_run respectively) so that the
+            # mode does not need to be passed to various methods, where a
+            # pair of such keys seem clearer than one "effective mode" key.
         }
+
+        effective_mode = mode  # actual mode to use for the first IO iteration
+        effective_fields = fields
+        if mode == "a":
+            # First read in the fields from the existing file:
+            effective_fields = NetCDFRead(_implementation).read(filename)
```

Comment:
Better to use self.implementation, i.e. effective_fields = NetCDFRead(self.implementation).read(filename), as this will play nicely with downstream applications (i.e. cf.python)
```suggestion
            effective_fields = NetCDFRead(self.implementation).read(filename)
```

Reply:
Excellent suggestion again, thanks. Not sure how I missed that since it is already used throughout the class!

I'll commit your suggestion here but also remove the next commit the explicit `implementation` import and variable definition I added previously related to this and they will no longer be required (as a bonus).